modules/storages/app/models/storages/storage.rb

```diff
@@ -147,6 +147,13 @@ def available_project_folder_modes
       raise Errors::SubclassResponsibility
     end
 
+    # Returns a value of an audience, if configured for this storage.
+    # The presence of an audience signals that this storage prioritizes
+    # remote authentication over token exchange if possible.
+    def audience
+      raise Errors::SubclassResponsibility
```

Comment:
The interface already defines that returning an audience means "I support SSO authentication" and not returning one means that no SSO authentication is supported.

What do you think about implementing the default behaviour already here as `nil` and documenting that subclasses that want to support SSO authentication need to return an audience?

This would effectively allow us to drop the empty implementation over at OneDrive and reduce the amount of boilercode required to implement a minimal storage.

Reply:
I see the thought behind it, yet I would disagree because of one detail.

In a perfect world, this one would be an abstract class and then default behaviour could be implemented here. Yet, in ruby you CAN instantiate a `storage`, so we need to "break" some of his methods.

There is another, even more important paradigm for this, and this is unintended functionality. If the base class implements default behaviour, this can shadow the situation, where a subclass functionality was never defined.

This is why I really like to raise errors on places in a base class where I want to force any developer - even me - to think about, how this needs to be implemented by each subclass.